### PR TITLE
DFPL-407 Sendgrid to send notification via email attachment of change…

### PR DESCRIPTION
**Pull request by so99y**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DFPL-407

### Change description ###
As Cafcass 

I want to receive an "email notification " for any change of address  by HMCTS users to [publiclawintake@cafcass.gov.uk](mailto:publiclawintake@cafcass.gov.uk). from the portal 

So that I can be kept updated of current changes

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: service/src/main/java/uk/gov/hmcts/reform/fpl/utils/PeopleInCaseHelper.java [hunk 1]
```diff
@@ -68,4 +74,29 @@ public static String getFirstRespondentFullName(List<Element<Respondent>> respon
...
+    }
+
+    public static boolean hasAddressChange(List<Element<? extends ConfidentialParty<?>>> after,
+                                           List<Element<? extends ConfidentialParty<?>>> before) {
+        Map<UUID, Address> beforeChildAddresses = addressByConfidentialPartyId(before);
+        String delimiter = ",";
+        for (Element<? extends ConfidentialParty> a : after) {
+            Address beforeAddress = beforeChildAddresses.get(a.getId());
+            if (beforeAddress != null) {
+                if (!equalsIgnoreCase(beforeAddress.getAddressAsString(delimiter),
```
Comment (SomeshRA) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/utils/PeopleInCaseHelper.java:94]: Wouldn't equals on Address objects work?

Comment (so99y) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/utils/PeopleInCaseHelper.java]: Alright.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/RespondentsUpdatedEventHandler.java [hunk 2]
```diff
@@ -27,6 +32,24 @@
...
+        CaseData caseData = event.getCaseData();
+        CaseData caseDataBefore = event.getCaseDataBefore();
+
+        final Optional<CafcassLookupConfiguration.Cafcass> recipientIsEngland =
+            cafcassLookupConfiguration.getCafcassEngland(caseData.getCaseLocalAuthority());
+
+        if (recipientIsEngland.isPresent() && respondentService.hasAddressChange(caseData.getAllRespondents(),
+            caseDataBefore.getAllRespondents())) {
+            cafcassNotificationService.sendEmail(caseData, CHANGE_OF_ADDRESS,
+                ChangeOfAddressData.builder().build());
```
Comment (SomeshRA) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/RespondentsUpdatedEventHandler.java:50]: Don't they need to know who's address had changed?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/ChildrenUpdatedEventHandler.java [hunk 3]
```diff
@@ -30,6 +36,25 @@
...
+        CaseData caseData = event.getCaseData();
+        CaseData caseDataBefore = event.getCaseDataBefore();
+
+        final Optional<CafcassLookupConfiguration.Cafcass> recipientIsEngland =
+            cafcassLookupConfiguration.getCafcassEngland(caseData.getCaseLocalAuthority());
+
+        if (recipientIsEngland.isPresent() && childrenService.hasAddressChange(caseData.getAllChildren(),
+            caseDataBefore.getAllChildren())) {
+            cafcassNotificationService.sendEmail(caseData, CHANGE_OF_ADDRESS,
+                ChangeOfAddressData.builder().build());
```
Comment (SomeshRA) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/handlers/ChildrenUpdatedEventHandler.java:55]: Don't they need to know who's address had changed?

**Review by SomeshRA — APPROVED** [review]

**Review by SomeshRA — DISMISSED** [review]

**Pull request merged**